src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,122 @@
+Here `𝕜[X]` is a PID, so there is a polynomial `g ∈ alg_hom.annihilating_ideal u`
+which generates the ideal. We show that if this generator is
+chosen to be monic, then it is the minimal polynomial of `u`,
+as defined in `field_theory.minpoly`.
+-/
+
+variables {A M B : Type*}
+
+variables [comm_ring A] [add_comm_group M] [module A M]
+
+open_locale polynomial
+
+open polynomial
+
+/--
+First we make sense of the informal notation `p(u)`.
+Think of `u` as the image of `X` in an algebra map `A[X] → module.End A M`
+and extend this to all polynomials `p(X)` using the `A` algebra structure
+on `module.End A M` (`= (M →ₗ[A] M)`).  This is given to us by `polynomial.aeval`.
+Using this definition, we can define the annihilating ideal of `u` to be
+`{ p ∈ A[x] | p(u) = 0 }`. This is an ideal in `A[X]`. -/
+noncomputable def annihilating_ideal (u : M →ₗ[A] M) : ideal A[X] :=
+ring_hom.ker (aeval u).to_ring_hom
+
+/-- It is useful to refer to ideal membership sometimes
+ and the annihilation condition other times -/
+lemma mem_annihilating_ideal_iff_aeval_0 (u : M →ₗ[A] M) (p : A[X]) :
+  p ∈ annihilating_ideal u ↔ aeval u p = 0 :=
+iff.rfl
+
+variables {𝕜 V : Type*}
+variables [field 𝕜] [add_comm_group V] [module 𝕜 V]
+
+/-- Since `𝕜[x]` is a principal ideal domain there is a polynomial `g` such that
+ `span 𝕜 {g} = annihilating_ideal u` -/
+noncomputable def annihilating_ideal_generator (u : V →ₗ[𝕜] V) : 𝕜[X] :=
+submodule.is_principal.generator (annihilating_ideal u)
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the first condition: it must annihilate
+the original endomorphism `u : V →ₗ[𝕜] V`. -/
+lemma annihilating_ideal_generator_aeval_0 (u : V →ₗ[𝕜] V) :
+  aeval u (annihilating_ideal_generator u) = 0 :=
+begin
+  rw annihilating_ideal_generator,
+  have gen_u_member := submodule.is_principal.generator_mem (annihilating_ideal u),
+  exact (ring_hom.mem_ker (polynomial.aeval u).to_ring_hom).1 gen_u_member,
+end
+
+/-- This is a stepping stone to show the generator has minimal degree -/
+lemma mem_iff_generator_dvd (u : V →ₗ[𝕜] V) (p : 𝕜[X]) :
+  p ∈ annihilating_ideal u ↔ annihilating_ideal_generator u ∣ p :=
+submodule.is_principal.mem_iff_generator_dvd (annihilating_ideal u)
+
+/-- We are working toward showing the generator of the annihilating ideal
+in the field case is the minimal polynomial. We are going to use a uniqueness
+theorem of the minimal polynomial. This is the second condition: it must have minimal
+degree among the annihilators of the original endomorphism `u : V →ₗ[𝕜] V`. -/
+lemma non_zero_gen_of_non_zero_mem (u : V →ₗ[𝕜] V) (p : 𝕜[X])
+  (hp : p ∈ annihilating_ideal u) (hpn0 : p ≠ 0) :
+  annihilating_ideal_generator u ≠ 0 :=
+begin
+  intro hg,
+  cases (mem_iff_generator_dvd u p).1 hp with q hq,
+  rw [ hg, zero_mul ] at hq,
+  exact hpn0 hq,
+end
```

Comment:
It's easier to prove the positive version of this first:
```lean
lemma eq_zero_of_mem_eq_zero (u : V →ₗ[𝕜] V) (p : 𝕜[X])
  (hp : p ∈ annihilating_ideal u) (hu0 : annihilating_ideal_generator u = 0) :
  p = 0 :=
begin
  rw mem_iff_generator_dvd at hp,
  rwa [hu0, zero_dvd_iff] at hp,
end


/-- We are working toward showing the generator of the annihilating ideal
in the field case is the minimal polynomial. We are going to use a uniqueness
theorem of the minimal polynomial. This is the second condition: it must have minimal
degree among the annihilators of the original endomorphism `u : V →ₗ[𝕜] V`. -/
lemma non_zero_gen_of_non_zero_mem (u : V →ₗ[𝕜] V) (p : 𝕜[X])
  (hp : p ∈ annihilating_ideal u) (hpn0 : p ≠ 0) :
  annihilating_ideal_generator u ≠ 0 :=
mt (eq_zero_of_mem_eq_zero u p hp) hpn0
```

Reply:
Turns out I was using this at one point, but after adjusting proofs according to suggestions given here, I no longer need this. Should I 
- remove it
- leave it but move it below the main proof since it is an ancillary fact
- leave it where it is, comment that it is just a simple fact that may be of use later
- other??

Reply:
@eric-wieser Aside from this issue, do you see any remaining items?  
Some things I can think of: Update the wiki to mark this as done (don't know how), Add more lemmas (not sure what would be useful).

Reply:
> me things I can think of: Update the wiki to mark this as done (don't know how), Add more lemmas (not sure what would be useful).

You need to edit the `undergrad.yaml` file here:

https://github.com/leanprover-community/mathlib/blob/e3d36811ea404ab6f009e3d5648741e144ff77ba/docs/undergrad.yaml#L54

to link to the definition you think is most relevant.